Add reading of qubit_lo_freq from pulse Schedule config

DanPuzzuoli opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

Closes #1179

Follows logic as suggested in #1179 by @nkanazawa1989 : a global `qubit_lo_freq` field is populated as before, either from the qobj config or from the Hamiltonian. When simulating a specific schedule, if no `qubit_lo_freq` is present in the schedule config, the global default is used, otherwise the schedule's `qubit_lo_freq` is used.


### Details and comments

When each schedule is parsed (in dict format in `experiment_to_struct`) it checks if it has a `'config'` key, and if so, will check for a `qubit_lo_freq` field. If it's there, this will be stored in the "experiment struct" dictionary, otherwise, the `struct['qubit_lo_freq']` is set to `None`. When setting up the RHS function before solving, if `struct['qubit_lo_freq'] is not None`, this value will be subbed in, otherwise the default is maintained.

A test has been added to `test_pulse_simulator` with the same syntax as the example given by @nkanazawa1989  in #1179. A schedule which should do a perfect pi pulse (if on resonance) is run on a model with 3 different `qubit_lo_freq` values: the first below resonance, the second on resonance, and the third above resonance. The test verifies that the two off-resonant schedules fail to excite the system by checking that all measurement counts are in the ground state, whereas the on-resonance schedule returns all counts in the excited state.

### Question

Pending CI tests, this PR should be done. One potential issue that could arise with the current code is to do with formatting of the `qubit_lo_freq` field in the schedule config. The pulse simulator stores the default frequencies as a list of floats, which I believe are assumed to be canonically ordered, first with all drive frequencies in ascending order, then with all control channels. The `qubit_lo_freq` in the schedule config seems to just be a list (i.e. it doesn't contain channel name information) - is it safe to assume that it is also ordered in this way, or is there some other way to figure out the order in this list?  @nkanazawa1989

nkanazawa1989 commented on qiskit/providers/aer/pulse/controllers/pulse_controller.py:
```diff
@@ -422,7 +423,15 @@ def init_rhs(self, exp):
+        rhs_dict = self._rhs_dict
+
+        # if the experiment overrides qubit_lo_freq, use this one
+        if exp.get('qubit_lo_freq', None) is not None:
```
why not just `if 'qubit_lo_freq' in exp:`?

DanPuzzuoli replied:
True :)

DanPuzzuoli replied:
Actually I'm still going to keep a check for whether it `is not None`

nkanazawa1989 reviewed (approved):
Thanks @DanPuzzuoli ! This looks good to me. Suffice to say that I'm bit worried about how much is the new unittest robust to numerical error because it looks like an integration test. I think this is still ok, but I would separate a function to generate the rhs dict and add a specific test for that.

As you mentioned qubit lo frequencies are provided by a list without any mapping to channel. We just assume straightforward mapping of qubit index to drive channel index. More precisely the actual map may be provided by the backend, i.e. `backend.configuration().qubit_channel_mapping`, but in the openpulse specification it is not allowed to have multiple drive channels per qubit. So the mapping is obvious. However in the simulator we can define arbitrary Hamiltonian, so this constraint may limit the capability of simulation. However I think this is outside the scope of this PR.

DanPuzzuoli commented:
@nkanazawa1989 

I've changed:
- Moved the frequency overriding in the RHS dict to its own function `setup_rhs_dict_freqs`. I realized as well that with the previous commit I had completely forgotten that the frequencies in schedule config would just be specifying the `DriveChannel` frequencies (so ordering isn't an issue), and that I needed to call the `calculate_channel_frequencies` method of the `system_model` to compute the control channel frequencies as well (which will put them in whatever order is required by the model, so that's good!). So, I've made `setup_rhs_dict_freqs` also take in a third argument `calculate_channel_frequencies` so that the logic can have access to this function.
- I've added a direct test of this function in `test.pulse.controllers.test_pulse_controller` - making sure that it works when there is no `qubit_lo_freq` key in the exp dictionary, and making sure it works when there is.
- Totally unrelated: I noticed there were some test files without copyright strings, so I added those.

nkanazawa1989 reviewed (approved):
Thanks for reviewing guide! I think this is good to go.
